parallel-install/pkg/preinstaller/preinstaller.go

```diff
@@ -0,0 +1,127 @@
+package preinstaller
+
+import (
+	"fmt"
+	"github.com/avast/retry-go"
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/config"
+	"io/ioutil"
+	"k8s.io/client-go/dynamic"
+	"os"
+)
+
+// PreInstaller prepares k8s cluster for Kyma installation.
+type PreInstaller struct {
+	applier       ResourceApplier
+	cfg           config.Config
+	dynamicClient dynamic.Interface
+	retryOptions  []retry.Option
+}
+
+// NewPreInstaller creates a new instance of PreInstaller.
+func NewPreInstaller(applier ResourceApplier, cfg config.Config, dynamicClient dynamic.Interface, retryOptions []retry.Option) *PreInstaller {
+	return &PreInstaller{
+		applier:       applier,
+		cfg:           cfg,
+		dynamicClient: dynamicClient,
+		retryOptions:  retryOptions,
+	}
+}
+
+// InstallCRDs on a k8s cluster.
+// Returns Output containing results of installation.
+func (i *PreInstaller) InstallCRDs() (Output, error) {
+	resource := resourceType{
+		name: "crds",
+	}
+
+	output, err := i.apply(resource)
+	if err != nil {
+		return Output{}, err
+	}
+
+	return output, nil
+}
+
+// CreateNamespaces in a k8s cluster.
+// Returns Output containing results of installation.
+func (i *PreInstaller) CreateNamespaces() (Output, error) {
+	resource := resourceType{
+		name: "namespaces",
+	}
+
+	output, err := i.apply(resource)
+	if err != nil {
+		return Output{}, err
+	}
+
+	return output, nil
+}
+
+func (i *PreInstaller) apply(resourceType resourceType) (o Output, err error) {
+	installationResourcePath := i.cfg.InstallationResourcePath
+	path := fmt.Sprintf("%s/%s", installationResourcePath, resourceType.name)
+
+	rawComponentsDir, err := ioutil.ReadDir(path)
+	if err != nil {
+		return o, err
+	}
+
+	components := findOnlyDirectoriesAmong(rawComponentsDir)
+	if components == nil || len(components) == 0 {
+		i.cfg.Log.Warn("There were no components detected for installation. Skipping.")
+		return o, nil
+	}
+
+	for _, component := range components {
+		componentName := component.Name()
+		i.cfg.Log.Infof("Processing component: %s", componentName)
```

Comment:
I think that context is missing here: we don't know which resourceType is being processed. So, if we have multiple resource types, we'll have few identical log entries: "Processing component: Foo", one for every resource type.